packages/editor/src/lib/editor/managers/SnapManager.ts

```diff
@@ -440,180 +457,92 @@ export class SnapManager {
-	// note: this happens within page space
-	snapLineHandleTranslate({
-		lineId,
-		handleId,
-		handlePoint,
-	}: {
-		lineId: TLShapeId
-		handleId: string
-		handlePoint: Vec2d
-	}): SnapData {
-		const line = this.editor.getShapeById<TLLineShape>(lineId)
-		if (!line) {
-			return { nudge: new Vec2d(0, 0) }
-		}
-
-		// We want the line to be able to snap to itself!
-		// but we don't want it to snap to the current segment we're drawing
-		// so let's get the splines of all segments except the current one
-		// and then pass them to the snap function as 'additionalOutlines'
-
-		// First, let's find which handle we're dragging
-		const util = this.editor.getShapeUtil(LineShapeUtil)
-		const handles = util.handles(line).sort(sortByIndex)
-		if (handles.length < 3) return { nudge: new Vec2d(0, 0) }
-
-		const handleNumber = handles.findIndex((h) => h.id === handleId)
-		const handle = handles[handleNumber]
-
-		// Now, let's figure out which segment this handle is on
-		// So... there are two types of handles:
-		// - vertex
-		// - create
-
-		// And this is how the handles of a line are arranged:
-		// vertex --- create --- vertex -- create -- vertex
-
-		// And we number them like this:
-		// v --- c --- v --- c --- v
-		// 0 --- 1 --- 2 --- 3 --- 4
-
-		// We want to get the segments made by connecting the vertex handles:
-		// v --- c --- v --- c --- v
-		// 0 --- 1 --- 2 --- 3 --- 4
-		// |-----------|-----------|
-		// | segment 0 | segment 1 |
-		// |-----------|-----------|
-
-		// If we're dragging a vertex handle, we can get its segment number by dividing its handle number by 2
-		// If we're dragging a create handle, we can get its segment number by adding 1 to its handle number, then dividing by 2
-		const segmentNumber = handle.type === 'vertex' ? handleNumber / 2 : (handleNumber + 1) / 2
-
-		// Then, get the splines of all segments except the current one
-		// (and by the way - we want to get the splines in page space, not shape space)
-		const spline = getSplineForLineShape(line)
-		const ignoreCount = 1
-		const pageTransform = this.editor.getPageTransform(line)!
-
-		const pageHeadSegments = spline.segments
-			.slice(0, Math.max(0, segmentNumber - ignoreCount))
-			.map((s) => Matrix2d.applyToPoints(pageTransform, s.lut))
-
-		const pageTailSegments = spline.segments
-			.slice(segmentNumber + ignoreCount)
-			.map((s) => Matrix2d.applyToPoints(pageTransform, s.lut))
-
-		return this.snapHandleTranslate({
-			handlePoint: handlePoint,
-			additionalOutlines: [...pageHeadSegments, ...pageTailSegments],
+	@computed get outlinesInPageSpace() {
+		return this.snappableShapes.map(({ id, isClosed }) => {
+			const outline = deepCopy(this.editor.getOutlineById(id))
+			if (isClosed) outline.push(outline[0])
+			const pageTransform = this.editor.getPageTransformById(id)
+			if (!pageTransform) throw Error('No page transform')
+			return Matrix2d.applyToPoints(pageTransform, outline)
```

Comment:
These are kind of expensive to compute so we do them once here rather than on every frame.